frontend/device/android/powerd.lua

```diff
@@ -2,18 +2,48 @@ local BasePowerD = require("device/generic/powerd")
 local _, android = pcall(require, "android")
 
 local AndroidPowerD = BasePowerD:new{
-    fl_min = 0, fl_max = 25,
-    fl_intensity = 10,
+    fl_min = 0, fl_max = 100, -- consistant scale across android, kobo ...
+--    fl_intensity = 10,
 }
 
 function AndroidPowerD:frontlightIntensityHW()
-    return math.floor(android.getScreenBrightness() / 255 * self.fl_max)
+    local intensity = math.floor(android.getScreenBrightness() / android.getScreenMaxBrightness() * self.fl_max)
+    if intensity < 0 then
+        intensity = 0
+    end
+    return intensity
 end
 
 function AndroidPowerD:setIntensityHW(intensity)
-    android.setScreenBrightness(math.floor(255 * intensity / self.fl_max))
+    self.fl_intensity = intensity
+    android.setScreenBrightness(math.floor(intensity * android.getScreenMaxBrightness() / self.fl_max))
 end
 
+function AndroidPowerD:init()
+    if self.device:hasNaturalLight() then
+        self.fl_warmth = self.getWarmth()
+    end
+end
+
+function AndroidPowerD:setWarmth(warmth)
+    self.fl_warmth = warmth
+    android.setScreenWarmth( warmth / AndroidPowerD:getMaxWarmth() )
+end
+
+function AndroidPowerD:getWarmth()
+    local warmth = android.getScreenWarmth() * AndroidPowerD.getMaxWarmth()
+    return warmth
+end
+
+function AndroidPowerD:getMaxWarmth()
```

Comment:
Probably AndroidPowerD:getMaxWarmth(),  AndroidPowerD:getMinWarmth() etc are not needed. We might want them initialized once. For now using maxWarmth is ok as minWarmth/minBrightness is always 0, but that doesn't need to apply on some circunstances.

So I would go for adding two ranges on the new instance, for example

```lua

local AndroidPowerD = BasePowerD:new{
    brightness_range = android.getScreenMaxBrightness() - android.getScreenMinBrightness(),
    warmth_range = android.getScreenMaxWarmth() - android.getScreenMinWarmth(),
}
```

Then you can use `self.brightness_range` instead of `AndroidPowerD:getMaxBrightness` and `self.warmth_range` instead of `AndroidPowerD:getMaxWarmth`

Reply:
We can do so, but then we have to scale the brightness to our brightness range AND add the minBrightness as an offset.

Reply:
???

Reply:
Lets say it different. The hardware has a range from `min` to `max`.  Min does not have to be 0.
The software has a range from [0..100]. 
If I want to map `value`  from the hardware scale to the software range I need `min` and `max`.
If I want to map `value` form the software range to the hardware scale I also need `min` and `max`.

So if I introduce a brightness_range, I would also need a brightness_offset. 
That`s why I am saying two variables.

Reply:
Ehm, nope.

You need to scale whatever difference between max and min values given by the android SDK to the scale used by the frontend (now 0 - 100). Probably `range` is not a good name for this variable, but that's it.

Reply:
Now I see. We proceed like you suggest. My view was biased by embedded programming.

In the future we should not forget: Our backend controller functions (setWarmth, setBrightness) expect values from 0 to Max_x. Our getXXXX functions return values from 0 to Max_x.
0 means dark or cold,  Max_x means their respective maximum brightness or warmth. That is the way we do now.
So the backend controller functions  (setWarmth, setBrightness) have to take care if the range is inverted (as we do now for warmth) and if a new super duper fancy device has its hardware minimum not at 0 then the setWarmth/setBrightness function shift the value.

If we do so, we can drop getMinWarmth and getMinBrightness as they are not needed.
Would that be a feasible solution?

Reply:
the backend code defines a interface but doesn't implement it. The interface allow us to encapsulate device specific implementations in a common way. Nor the generic impl or the Tolino impl do what you said: they check that a given value is between min and max before setting that value (for warmth and brightness) and they return implementation min/max values. It happens that in all the current implementations min = 0 but that's not part of the interface, just a specific implementation of it.

Reply:
> the backend code defines a interface but doesn't implement it. The interface allow us to encapsulate device specific implementations in a common way.

OK with that.
> Nor the generic impl or the Tolino impl do what you said: they check that a given value is between min and max before setting that value (for warmth and brightness)

The warmth interface inverts the range. On the /sys/class../color file 0 means warm and 10 means cold. In line 51 of TolinoWarmthController there is  `WARMTH_MAX - colorFile.readText().toInt()`

> and they return implementation min/max values. It happens that in all the current implementations min = 0 but that's not part of the interface, just a specific implementation of it.

Why shouldn't we make it part of the interface? Up to now it would make no difference.
Again: If we don`t make it part of the interface then we would need 2 values an offset and a range.

Reply:
> The warmth interface inverts the range.

Nope. The **Tolino implementation** of the interface inverts the range. The interface is generic enough and doesn't care about specific implementations.

> Why shouldn't we make it part of the interface? 

Please do read https://www.tutorialspoint.com/java/java_interfaces.htm.  I can't solve your issues if you don't try to understand what an interface is :p.

Reply:
> The warmth interface inverts the range. On the /sys/class../color file 0 means warm and 10 means cold. In line 51 of TolinoWarmthController there is WARMTH_MAX - colorFile.readText().toInt()

Historical quirk. The first backend didn't. Newer ones does (hence nl_invert on Kobos w/ a mixer).

Reply:
Thank you for your patience. The whole discussion is about mapping a value from one range to another as for example is described in https://www.arduino.cc/reference/en/language/functions/math/map/

Somewhere it has to be done! For the mapping on every side two values are necessary. In our case it can be assumed one of the four parameters is zero. So we need two ranges and one offset. 

I have the feeling we are not talking the same language. So I give you the code. Do whaterver you want with it, I won't work on it anymore.

My first PR (https://github.com/koreader/koreader/pull/6332) was not worse than the actual one. IMHO the user experience in KOReader was more consistent with the former one.

I am out now!

Reply:
@zwim: please do whatever you want but don't take it personally. As I said your first PR was ok as a patch for your own usage. It will crash on @elvvis device as he didn't have root and JNI calls are not protected.

This whole PR is almost done. We don't speak the same mother tongue (I speak galician and you speak german)  but we're both almost good with english (and have google translate to workaround our limitations :))

> I won't work on it anymore.

That's fine too but I'm afraid I cannot test the implementation myself, so I will wait for any tester before merging

> The whole discussion is about mapping a value from one range to another

That can be implemented on the backend or on the frontend, but has nothing to do with the interface itself, just with the implementation. Feel free to code it as you want.

> I am out now!

Closing this issue for the moment. Please ping me (or any other person here) if you want to open it again. AFAICT there's nothing wrong with current implementation, except for https://github.com/koreader/koreader/pull/6426#issuecomment-663864729, but that has nothing to do with your code.

Reply:
@pazos: 
First, the most important thing: It is nothing personal. You did a good job. It was fine to work with you. My saying about the language has nothing to do with our mother tongue,  I meant it abstract, we do not understand each other. Maybe this is due to the medium we are talking. In real life it would be different. You know what you are doing, so do I.

Second thing: If you need testers, feel free to give me a link to an apk.

Third thing: I could be implemented in  devicelistener.lua line 204 and 113 by calling checking if getXXXX is the same than powerd.fl_intensity, but poire-z's solution seems more elegant.

Fourth thing: I will look on the changes in the showcase tomorrow.

Reply:
Ok, thank you. This conversation was resolved successfully 👍. I will comment the third thing below.